extensions/amp-instagram/1.0/instagram.js

```diff
@@ -0,0 +1,119 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Preact from '../../../src/preact';
+import {ContainWrapper} from '../../../src/preact/component';
+import {dict} from '../../../src/utils/object';
+import {getData} from '../../../src/event-helper';
+import {parseJson} from '../../../src/json';
+import {useLayoutEffect, useRef, useState} from '../../../src/preact';
+
+const NO_HEIGHT_STYLE = dict();
+
+/**
+ * @param {!InstagramPropsDef} props
+ * @return {PreactDef.Renderable}
+ */
+export function Instagram({
+  shortcode,
+  captioned,
+  title,
+  requestResize,
+  ...rest
+}) {
+  const iframeRef = useRef(null);
+  const [heightStyle, setHeightStyle] = useState(NO_HEIGHT_STYLE);
+  const [opacity, setOpacity] = useState(0);
+
+  useLayoutEffect(() => {
+    const messageHandler = (event) =>
+      handleMessage(
+        event,
+        iframeRef,
+        requestResize,
+        setHeightStyle,
+        setOpacity
+      );
+    const {defaultView} = iframeRef.current.ownerDocument;
+
+    defaultView.addEventListener('message', messageHandler);
+
+    return () => {
+      defaultView.removeEventListener('message', messageHandler);
+    };
+  }, [requestResize]);
+
+  return (
+    <ContainWrapper {...rest} wrapperStyle={heightStyle} layout size paint>
+      <iframe
+        ref={iframeRef}
+        src={
+          'https://www.instagram.com/p/' +
+          encodeURIComponent(shortcode) +
+          '/embed/' +
+          (captioned ? 'captioned/' : '') +
+          '?cr=1&v=12'
+        }
+        scrolling="no"
+        frameborder="0"
+        allowtransparency
+        title={title}
+        loading="lazy"
+        style={{
+          width: '100%',
+          height: '100%',
+          opacity,
+          contentVisibility: 'auto',
+        }}
+      />
+    </ContainWrapper>
+  );
+}
+
+/**
+ * @param {Event} event
+ * @param {Object} iframeRef
+ * @param {function(number):*|undefined} requestResize
+ * @param {function(Object)} setHeightStyle
+ * @param {function(number)} setOpacity
+ * @return {undefined}
+ */
+export function handleMessage(
```

Comment:
Why is this exported? Only for testing? Is there a way to test w/o exposing it? If not, let's add `@visibleForTesting`. But, IMHO, it'd be much more preferred if we could test w/o exposing it directly.

Reply:
Yeah, its exported and visible for testing. I tried various ways of testing the iframe's layout callback without exporting it, but the iframe wasn't able to properly receive any mock event sent. This would either lead to timeout errors (for not receiving it in time) or errors for not receiving the message. Since that method was pretty flaky, I settled on this method to properly test the component consistently.

Reply:
Iframe shouldn't be receiving any events, should it? Only the parent window receives the events? Is it not possible to just call:

```
const mockEvent = new Event('message');
mockEvent.origin = '...';
mockEvent.source = iframe.contentWindow;
win.dispatchEvent(mockEvent);
```

?

Reply:
Yeah, I tried this out and it leads to a lot of flakiness. Sometimes the event is received, sometimes it isn't. I'd say locally the test passes 50% of the time. 

`const messageHandler = (event) =>
      handleMessage(
        event,
        iframeRef,
        requestResize,
        setHeightStyle,
        setOpacity
      );`

I can still change it but I'm not sure why it's not consistent.

Reply:
That's exactly the kind of thing a test could help a lot with to clarify. Different bugs related to race conditions are very possible here. I'd start with a simple debug session. Can you trace that a message is deliverable and it's just flaky? If so, we should try to understand and test this flakiness source.

Reply:
The message is deliverable. I'm converting the PR to a draft and uploading the code to see if maybe it's a local issue. I know Travis is more consistent.

Reply:
The test that fails is the "height change" test, where the height being changed isn't being recorded by the `expect` of the test. After debugging, it looks like this is a possible race condition or testing issue, where the testing framework isn't able to recognize the change since the event processing happens slower than the check.

Reply:
Update: It looks like it was officially a race condition. The mockEvent is dispatched fast enough to be received, but the Instagram component isn't able to update the `ContainWrapper`'s `div` height fast enough for the test to verify it. As a result, I implemented a `waitForHeight` function that waits for the height to equal what we want it for the test to be passed. If for some reason the height is never set, the test will time out and throw an error. Hopefully this solves the issue with the test, allows us to keep the functions together without any refactoring, and possibly might be a useful pattern or util to have for future tests of other Bento components.